### PR TITLE
Add package wmderland

### PR DESCRIPTION
**Pull request by takagiy**
###### Motivation for this change
This request is going to add [wmderland](https://github.com/aesophor/wmderland), the X11 tiling window manager to nixpkgs and also some configuration of window managers for wmderland to nixos.

list of packages which will be available:
* `wmderland`: An X11 tiling window manager
* `wmderlandc`: An ipc-client to communicate with wmderland

list of configurations which will be available (in nixos):
* `services.xserver.windowManager.wmderland.enable`
        If `true`, wmderland is made available.
* `services.xserver.windowManager.wmderland.extraSessionCommands`
        Shell commands executed just before wmderland is started.
* `services.xserver.windowManager.wmderland.extraPackages`
        Extra packages to be installed system wide. (defaulted to `[ rofi dunst light hsetroot feh rxvt-unicode ]`)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
```console
$ nix path-info -S /run/current-system/sw/bin/wmderland
/nix/store/i956jcaqzlkwf4nfffxriw4gqlvvkzan-wmderland-2020-07-17	   43616424
$ nix path-info -S /run/current-system/sw/bin/wmderlandc
/nix/store/a9223wa17zsiakfmccyliwy1v8pb2rrb-wmderlandc-2020-07-17	   37202000
```
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    libX11
+    xorgproto
+  ];
+
+  meta = {
+    description = "Modern and minimal X11 tiling window manager";
+    homepage = "https://github.com/aesophor/wmderland";
+    license = licenses.mit;
+    platforms = libX11.meta.platforms;
+    maintainers = [ maintainers.takagiy ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/default.nix:32]:
```suggestion
    maintainers = with maintainers; [ takagiy ];
```

File: pkgs/applications/window-managers/wmderland/ipc-client.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    libX11
+    xorgproto
+  ];
+
+  meta = {
+    description = "A tiny program to interact with wmderland";
+    homepage = "https://github.com/aesophor/wmderland/tree/master/ipc-client";
+    license = licenses.mit;
+    platforms = platforms.all;
+    maintainers = [ maintainers.takagiy ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/ipc-client.nix:32]:
```suggestion
    maintainers = with maintainers; [ takagiy ];
```

**Comment by SuperSandro2000** [comment]
Can you squash the fix up commits together?

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -8403,6 +8403,12 @@
     githubId = 321799;
     name = "Paul Colomiets";
   };
+  takagiy = {
+    email = "takagiy.4dev@gmail.com";
+    github = "takagiy";
+    githubID = "18656090";
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:8409]:
```suggestion
    githubID = 18656090;
```

**Comment by takagiy** [comment]
typo. (githubID -> githubId) I will fix it.
https://github.com/NixOS/nixpkgs/blob/e28dcc4d54f92d54ba8c2e68f0a4741786bc64d8/maintainers/maintainer-list.nix#L8409

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -24191,6 +24191,10 @@ in
 
   wmctrl = callPackage ../tools/X11/wmctrl { };
 
+  wmderland = callPackage ../applications/window-managers/wmderland { };
+
+  wmderlandc = callPackage ../applications/window-managers/wmderland/ipc-client.nix {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:24196]:
```suggestion
  wmderlandc = callPackage ../applications/window-managers/wmderlandc {};
```

The files don't share anything major so we should do this.

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
+{ stdenv, fetchFromGitHub, cmake, libX11, xorgproto }:
+
+with stdenv.lib;
+
+stdenv.mkDerivation {
+  pname = "wmderland";
+  version = "2020-07-17";
```
Comment (sikmir) [inline pkgs/applications/window-managers/wmderland/default.nix:7]: Is there any reasons why not using latest release 1.0.4?

Comment (takagiy) [reply pkgs/applications/window-managers/wmderland/default.nix:7]: It seems that v1.0.4 is so behind from the latest commit and some major features already implemented (e.g. varying split ratio) is missing.

Comment (sikmir) [reply pkgs/applications/window-managers/wmderland/default.nix:7]: Well, add `unstable-` prefix to version at least.

Comment (takagiy) [reply pkgs/applications/window-managers/wmderland/default.nix]: Thnak you. I've added it.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (takagiy) [reply pkgs/top-level/all-packages.nix]: > The files don't share anything major so we should do this. This would also require to move and rename the nix file.

Thank you. I've fixed it.

> Also wmderlandc builds on darwin but wanderland not:

Thank you again for your reviewing.
I finally found that that errors caused by the link-time optimization flag (`-flto`) defined in CMakeLists.txt. (https://github.com/aesophor/wmderland/blob/a40a3505dd735b401d937203ab6d8d1978307d72/CMakeLists.txt#L13)

Hence, I must remove this linker option when the package is built on darwin. But I am not so clear how it's done. Could you give me an idea?

Comment (takagiy) [reply pkgs/top-level/all-packages.nix]: I think that one possible solution is to write a patch to remove that option and to fetch (or placing the patch in the nixpkgs repository and loading it directly) and to apply it when the building platforms matches with darwin. How do you feel about it?

**Review by berbiche — CHANGES_REQUESTED** [review]
1. > Before starting wmderland, EDIT your config first! Or you'll get a blackscreen

    Copying the example configuration file to `$out/share/wmderland/config` would be a good idea.

2. A simple automated test can be added.
    See `nixos/tests/i3wm.nix` for a sample test.
    ```
    { stdenv, ..., nixosTests }:
    {
      pname = "wmderland";
      ...
      passthru.tests.basic = nixosTests.wmderland;
    }
    ```

3.
    There is a sample desktop session file in the repository.

    Copying this file to `$out/share/xsessions` would allow login managers to detect the WM without having to `wmderland.enable = true;`

    This would also need `passthru.providedSessions = [ "wmderland" ];`

**Comment by berbiche** [comment]
Additionally, `src/utils.cc` needs to have its call to `notify-send` replaced with `${libnotify}/bin/notify-send`
See https://github.com/aesophor/wmderland/blob/a40a3505dd735b401d937203ab6d8d1978307d72/src/util.cc#L265-L267

**Comment by berbiche** [comment]
<details>
<summary>Quick patch file if you are interested</summary>

``` patch
From adefee1cf35a39a0aa29b880720383162b60c364 Mon Sep 17 00:00:00 2001
From: Nicolas Berbiche <nicolas@normie.dev>
Date: Thu, 29 Oct 2020 20:32:05 -0400
Subject: [PATCH] wmderland: add tests and expose xsessions desktop

---
 nixos/tests/all-tests.nix                     |  1 +
 nixos/tests/wmderland.nix                     | 54 +++++++++++++++++++
 .../window-managers/wmderland/default.nix     | 12 ++++-
 3 files changed, 66 insertions(+), 1 deletion(-)
 create mode 100644 nixos/tests/wmderland.nix

diff --git a/nixos/tests/all-tests.nix b/nixos/tests/all-tests.nix
index 9ffeba27a7f..536778467df 100644
--- a/nixos/tests/all-tests.nix
+++ b/nixos/tests/all-tests.nix
@@ -371,6 +371,7 @@ in
   virtualbox = handleTestOn ["x86_64-linux"] ./virtualbox.nix {};
   wasabibackend = handleTest ./wasabibackend.nix {};
   wireguard = handleTest ./wireguard {};
+  wmderland = handleTest ./wmderland.nix {};
   wordpress = handleTest ./wordpress.nix {};
   xandikos = handleTest ./xandikos.nix {};
   xautolock = handleTest ./xautolock.nix {};
diff --git a/nixos/tests/wmderland.nix b/nixos/tests/wmderland.nix
new file mode 100644
index 00000000000..0129c1caa4b
--- /dev/null
+++ b/nixos/tests/wmderland.nix
@@ -0,0 +1,54 @@
+import ./make-test-python.nix ({ pkgs, ...} : {
+  name = "wmderland";
+  meta = with pkgs.stdenv.lib.maintainers; {
+    maintainers = [ berbiche ];
+  };
+
+  machine = { lib, ... }: {
+    imports = [ ./common/x11.nix ./common/user-account.nix ];
+    test-support.displayManager.auto.user = "alice";
+    services.xserver.displayManager.defaultSession = lib.mkForce "none+wmderland";
+    services.xserver.windowManager.wmderland.enable = true;
+
+    systemd.services.setupWmderlandConfig = {
+      wantedBy = [ "multi-user.target" ];
+      before = [ "multi-user.target" ];
+      environment = {
+        HOME = "/home/alice";
+      };
+      unitConfig = {
+        type = "oneshot";
+        RemainAfterExit = true;
+        user = "alice";
+      };
+      script = let
+        config = pkgs.writeText "config" ''
+             set $Mod = Mod1
+             bindsym $Mod+Return exec ${pkgs.xterm}/bin/xterm -cm -pc
+        '';
+      in ''
+        mkdir -p $HOME/.config/wmderland
+        cp ${config} $HOME/.config/wmderland/config
+      '';
+    };
+  };
+
+  testScript = { ... }: ''
+    with subtest("ensure x starts"):
+        machine.wait_for_x()
+        machine.wait_for_file("/home/alice/.Xauthority")
+        machine.succeed("xauth merge ~alice/.Xauthority")
+
+    with subtest("ensure we can open a new terminal"):
+        machine.send_key("alt-ret")
+        machine.wait_until_succeeds("pgrep xterm")
+        machine.wait_for_window(r"alice.*?machine")
+        machine.screenshot("terminal")
+
+    with subtest("ensure we can communicate through ipc with wmderlandc"):
+        # Kills the previously open xterm
+        machine.succeed("pgrep xterm")
+        machine.execute("DISPLAY=:0 wmderlandc kill")
+        machine.fail("pgrep xterm")
+  '';
+})
diff --git a/pkgs/applications/window-managers/wmderland/default.nix b/pkgs/applications/window-managers/wmderland/default.nix
index bc6341a5889..3098dd22bb5 100644
--- a/pkgs/applications/window-managers/wmderland/default.nix
+++ b/pkgs/applications/window-managers/wmderland/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, cmake, libX11, xorgproto }:
+{ stdenv, fetchFromGitHub, cmake, libX11, xorgproto, nixosTests }:
 
 with stdenv.lib;
 
@@ -26,6 +26,16 @@ stdenv.mkDerivation {
     xorgproto
   ];
 
+  postInstall = ''
+    install -Dm0644 -t $out/share/wmderland/contrib $src/example/config
+    install -Dm0644 -t $out/share/xsessions $src/example/wmderland.desktop
+  '';
+
+  passthru = {
+    tests.basic = nixosTests.wmderland;
+    providedSessions = [ "wmderland" ];
+  };
+
   meta = {
     description = "Modern and minimal X11 tiling window manager";
     homepage = "https://github.com/aesophor/wmderland";
--
2.28.0
```

</details>

File: nixos/modules/services/x11/window-managers/wmderland.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  cfg = config.services.xserver.windowManager.wmderland;
+in
+
+{
+  options.services.xserver.windowManager.wmderland = {
+    enable = mkEnableOption "wmderland";
+
+    extraPackages = mkOption {
+      type = with types; listOf package;
+      default = with pkgs; [ rofi ];
```
[the remark this answers is not in the document]
Comment (takagiy) [reply nixos/modules/services/x11/window-managers/wmderland.nix:15]: Thank you.
I was not sure what default value of this option should be. I’ll add some packages mentioned in example/config (something like rofi, urxvt, light and dust), and feh and hsetroot too (probably needed to set the desktop background).
Maybe, should some lock things be present? The original config uses [displayctl](https://github.com/aesophor/displayctl), an wrapper of i3lock-color, etc., which is, however, not in nixpkgs.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 101852` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>wmderland</li>
    <li>wmderlandc</li>
  </ul>
</details>

File: nixos/modules/services/x11/window-managers/wmderland.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+{
+  options.services.xserver.windowManager.wmderland = {
+    enable = mkEnableOption "wmderland";
+
+    extraSessionCommands = mkOption {
+      default = "";
+      type = types.lines;
+      description = ''
+        Sehll commands executed just before wmderland is started.
```
Comment (berbiche) [inline nixos/modules/services/x11/window-managers/wmderland.nix:17]:
```suggestion
        Shell commands executed just before wmderland is started.
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 101852` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>wmderland</li>
    <li>wmderlandc</li>
  </ul>
</details>

**Comment by takagiy** [comment]
@SuperSandro2000 Well..., still unmerged. What am I supposed to do?

File: maintainers/maintainer-list.nix [hunk 8]
```diff
@@ -8403,6 +8403,12 @@
     githubId = 321799;
     name = "Paul Colomiets";
   };
+  takagiy = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:8406]: Please put this into a separate commit with the message ``maintainers: add takagiy``.

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{ stdenv, fetchFromGitHub, cmake, libnotify, libX11, xorgproto, nixosTests }:
+
+with stdenv.lib;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/default.nix:3]: Stdenv.lib is deprecated as per recent contributing guide change. Also please only use with in the smallest possible scope.

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    sha256 = "0npmlnybblp82mfpinjbz7dhwqgpdqc1s63wc1zs8mlcs19pdh98";
+  };
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  cmakeBuildType = "MinSizeRel";
+
+  patches = optional stdenv.isDarwin ./0001-remove-flto.patch;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/default.nix:22]: I prefer it if we applied this patch everywhere to prevent bit rot.

File: pkgs/applications/window-managers/wmderlandc/default.nix [hunk 11]
```diff
@@ -0,0 +1,34 @@
+{ stdenv, fetchFromGitHub, cmake, libX11, xorgproto }:
+
+with stdenv.lib;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderlandc/default.nix:3]: Same as above.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Sorry, you slipped through my to-do list. Sadly on the meantime we got a contributing guide change you need to address.

**Comment by takagiy** [comment]
@SuperSandro2000 Updated. Thank you for reviewing!

File: pkgs/applications/window-managers/wmderlandc/default.nix [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    xorgproto
+  ];
+
+  meta = {
+    description = "A tiny program to interact with wmderland";
+    homepage = "https://github.com/aesophor/wmderland/tree/master/ipc-client";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.all;
+    maintainers = with lib.maintainers; [ takagiy ];
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderlandc/default.nix:31]:
```suggestion
  meta = with lib; {
    description = "A tiny program to interact with wmderland";
    homepage = "https://github.com/aesophor/wmderland/tree/master/ipc-client";
    license = licenses.mit;
    platforms = platforms.all;
    maintainers = with maintainers; [ takagiy ];
  };
```

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  };
+
+  meta = {
+    description = "Modern and minimal X11 tiling window manager";
+    homepage = "https://github.com/aesophor/wmderland";
+    license = lib.licenses.mit;
+    platforms = libX11.meta.platforms;
+    maintainers = with lib.maintainers; [ takagiy ];
+  };
+}
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/default.nix:49]:
```suggestion
  meta = with lib; {
    description = "Modern and minimal X11 tiling window manager";
    homepage = "https://github.com/aesophor/wmderland";
    license = licenses.mit;
    platforms = libX11.meta.platforms;
    maintainers = with maintainers; [ takagiy ];
  };
}
```

**Comment by takagiy** [comment]
@SuperSandro2000 Applied. Thank you, again. Would you prefer me to squash them?

**Comment by SuperSandro2000** [comment]
> Would you prefer me to squash them?

One commit for the maintainer, package and module please.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101852` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>wmderland</li>
<li>wmderlandc</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101852` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>wmderland</li>
<li>wmderlandc</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101852` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>wmderland</li>
<li>wmderlandc</li>
</ul>
</details>

**Comment by takagiy** [comment]
Squashed. Thank you.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101852` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>wmderland</li>
<li>wmderlandc</li>
</ul>
</details>

File: pkgs/applications/window-managers/wmderland/default.nix [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    sha256 = "0npmlnybblp82mfpinjbz7dhwqgpdqc1s63wc1zs8mlcs19pdh98";
+  };
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  cmakeBuildType = "MinSizeRel";
+
+  patches = ./0001-remove-flto.patch;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/default.nix:20]:
```suggestion
  patches = [ ./0001-remove-flto.patch ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/window-managers/wmderland/0001-remove-flto.patch [hunk 15]
```diff
@@ -0,0 +1,13 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 17a4944..33406f3 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -10,7 +10,7 @@ include(BuildType)
+ # Request C++14 standard, using new CMake variables.
+ set(CMAKE_CXX_STANDARD 14)
+ set(CMAKE_CXX_STANDARD_REQUIRED True)
+-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -flto -Wall")
++set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wmderland/0001-remove-flto.patch:10]: We could have used substituteInPlace for this but you don't need to change it.

**Review by SuperSandro2000 — APPROVED** [review]
packages look good

**Comment by berbiche** [comment]
@SuperSandro2000 @takagiy friendly ping

**Comment by SuperSandro2000** [comment]
@ofborg test wmderland

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by berbiche** [comment]
@SuperSandro2000 Thanks for merging this :tada:
